### ISSUE TITLE
Most Gaussians collapse into needle-like shapes

### ISSUE THREAD

**Issue by justin4ai** [jonathsch/becominglit#3]
Dear jonathsch,

Hello, I am reproducing the results of BecomingLit by using the exact same code of your latest push. The only changes I've made are 1) to hold four light patterns out from the training data for testing following [your last comment](https://github.com/jonathsch/becominglit/issues/1) and 2) to include EXP-1 to train.

Seeing your [demo](https://jonathsch.github.io/becominglit/), hard-edged visualized gaussians are properly constructed in BecomingLit:

<img width="897" height="495" alt="Image" src="https://github.com/user-attachments/assets/272d4aca-f5b9-44de-926f-92a699d7ab46" />

. However, when I tried to reproduce the same hard-edge visualization of gaussians, they show very needle-like shapes, which look definitely weird.

This is one of 250k trained results:
<img width="1716" height="871" alt="Image" src="https://github.com/user-attachments/assets/89c2c324-5354-49bb-a690-7ed61d646a25" />
<img width="489" height="750" alt="Image" src="https://github.com/user-attachments/assets/37967e33-3ab9-4494-b7d1-9217748b6502" />
.

And the following is 7kth and 15kth steps, respectively:

<img width="538" height="796" alt="Image" src="https://github.com/user-attachments/assets/aaf476ea-9f64-47c6-87e0-7a74dc395edd" />
<img width="538" height="793" alt="Image" src="https://github.com/user-attachments/assets/1c7292d2-010d-40ca-988b-853c9c6d4b0b" />.

If you have any idea of this happening, I'd be so glad if you let me know the reason or your experience, or possible causes.
Thanks for your help again!

Best,
Junyeong Ahn

**Comment by jonathsch** [comment]
Hi,

Based on my experience, when training converges to unusual outcomes, it is usually due to tracking failures. This can occur in sequences with significant head motion where most cameras capture the head from an extreme angle, resulting in failures in facial landmark detection. Could you please confirm whether the FLAME tracking appears reasonable for all training sequences? If it does, and you are still encountering issues, I would be happy to investigate further.

**Comment by justin4ai** [comment]
@jonathsch Thanks for your quick response! I also suspected FLAME tracking error and made sure to exclude EXP-1 for training - I think the fact your codebase also excludes it by default verifies it - which has some obvious error cases with large head pose changes. But still there are converging to weird gaussians although they are okay in image level).

<img width="344" height="415" alt="Image" src="https://github.com/user-attachments/assets/c8dee66d-6841-430e-a03e-a33696e8d2f3" />

I've gone through all tracking results per sequence and verified there are no significant errors for other sequences except EXP-1.  

**EXP-1 (Excluded)**

<img width="1071" height="517" alt="Image" src="https://github.com/user-attachments/assets/cc9729d8-b3db-43b3-a7eb-20e8e94e8971" />

**EXP-2**

<img width="1071" height="516" alt="Image" src="https://github.com/user-attachments/assets/f537c9cb-27a1-4a9b-a123-ae10d212cce9" />

**FREE**

<img width="1071" height="515" alt="Image" src="https://github.com/user-attachments/assets/1131f47b-a37e-4c13-90c6-7d9767c8a53b" />

**SEN-{1, 2, 3}**

<img width="1070" height="518" alt="Image" src="https://github.com/user-attachments/assets/298bfb5a-99aa-4ed6-8eae-f4a357732673" />

Would you please check if the trained model from your latest commit shows normal shapes of gaussians when visualizing?


Best,
Junyeong Ahn

**Comment by jonathsch** [comment]
Hi Junyeong,

Sorry for the late reply. Could you send me your full training config, ie, the `config.yaml` file that is written to the run folder at the beginning of training?

However, when I look at the renderings and error maps, everything else looks pretty reasonable on your side.

**Comment by justin4ai** [comment]
@jonathsch 

This is the ```config.yaml``` in the **run folder**:
subject: 1022
data:
  class_name: becominglit.data.bl_dataset.BecomingLitDataset
  subject: ${subject}
  sequences_subset:
  - EXP-2
  - EMOTIONS
  - SEN-1
  - SEN-2
  - SEN-3
  downscale_factor: 2
  black_level_subtraction: 4.0
  fully_lit_only: false
  apply_alpha: true
  length_unit: mm
val_data:
  class_name: becominglit.data.bl_dataset.BecomingLitDataset
  subject: ${subject}
  sequences_subset:
  - FREE
  downscale_factor: ${data.downscale_factor}
  black_level_subtraction: ${data.black_level_subtraction}
  fully_lit_only: ${data.fully_lit_only}
  apply_alpha: ${data.apply_alpha}
  length_unit: ${data.length_unit}
  cameras_subset:
  - '222200037'
dataloader:
  batch_size: 4
  num_workers: 8
  pin_memory: true
  persistent_workers: false
  shuffle: true
model_name: becominglit_release
model:
  class_name: becominglit.models.becominglit.BecomingLitModel
  diff_sh_deg: 6
  flame_config:
    shape_params: 300
    expr_params: 100
lr: 0.0005
optimizer:
  class_name: torch.optim.Adam
  per_module:
    geo_mod:
      lr: ${lr}
    spec_mod:
      lr: ${lr}
    expr_enc:
      lr: ${lr}
    view_enc:
      lr: ${lr}
    bsdf_mod:
      lr: ${lr}
    means:
      lr: 0.01
    albedo:
      lr: 0.002
    roughness:
      lr: 0.001
loss:
  class_name: becominglit.criterion.ModularLoss
  losses:
    rgb_l1:
      weight: 0.8
    rgb_ssim:
      weight: 0.2
    l2_reg:
      loss_fn: l2_reg
      key: normal_offsets
      schedule:
        class_name: becominglit.criterion.MonotonicWeightSchedule
        start: 0
        end: 100000
        init_value: 1.0
        target_value: 0.0
    scale_bounds:
      loss_fn: bound_primscale_rgca
      weight: 0.01
      min_scale: 0.1
      max_scale: 20.0
    negcolor:
      loss_fn: negcolor
      weight: 0.1
    alpha_loss:
      loss_fn: rgb_l2
      weight: 0.01
      src_key: alpha
      tgt_key: alpha
train:
  run_id: 1022_main_down2
  tag: ${model_name}.${subject}
  log_dir: /home/nas1/junyeongahn/becominglit_tmp/exp
  run_dir: ${.log_dir}/${.tag}/${.run_id}
  ckpt_dir: ${.run_dir}/checkpoints
  n_max_iters: 250000
  n_warmup_iters: 1000
  log_every_n_steps: 100
  summary_every_n_steps: 5000
  ckpt_every_n_steps: 50000
  validate_every_n_steps: 50000
summary:
  class_name: becominglit.models.becominglit.BecomingLitSummary


. 

And the visualized splats look weird again.

<img width="1483" height="666" alt="Image" src="https://github.com/user-attachments/assets/912e29c3-d1b4-4f20-9666-6a45a4c1b20e" />

The tracking is also fine for id 1022 just like what I've shown.

Also I trained becominglit on goliath dataset but still the problem exists. So I think it is just an inherent trait of becominglit and would be so glad if you visualize yours after training from scratch so that the code is all right!

<img width="653" height="502" alt="Image" src="https://github.com/user-attachments/assets/dec7fd25-cf52-4fd8-b312-6400d9a5b2aa" />

<img width="402" height="624" alt="Image" src="https://github.com/user-attachments/assets/cd18e7a1-5091-4491-9918-9e584a124517" />

Best,
Junyeong Ahn

**Comment by jonathsch** [comment]
I ran the training on our official dataset release, with the latest released code version, and using your config. On my side, the Gaussian primitives look fine, and I cannot reproduce the artifacts you observe.

<img width="2206" height="1608" alt="Image" src="https://github.com/user-attachments/assets/81c7ae64-f433-4859-bf0f-675c1742e867" />

**Comment by justin4ai** [comment]
Thanks for your help although the re-training from scratch must have been very exhaustive. I'd like to ask if you have any filtering of gaussians to be visualized? Or are they fully visualized?

I'm so glad and thankful you've been helping me with several issues. I appreciate you spare your precious time!

Best Regards,
Junyeong Ahn

**Comment by jonathsch** [comment]
The visualization shows the full set of Gaussians. If it helps you, I'm also happy to share the checkpoint with you.

**Comment by justin4ai** [comment]
Hello, they must be very helpful for me to figure out the causes! I'd be very happy to be shared through ```justinahn@kaist.ac.kr```. If there are a couple of checkpoints or the zipped codebase folder (since the git code is possibly imperfect in that I'm observing the same phenomenon even with goliath dataset), they all will be very helpful since I have all id set of dataset downloaded. Of course if things allow. Even single checkpoint will be very appreciated!

Thanks you so much for your help and I'd like to share the exact cause or solution for this problem here after I figure out.

Best Regards,
Junyeong Ahn

**Comment by justin4ai** [comment]
Hello, I sent you an email with details I've figured out with your checkpoint. In short, the statistics of gaussians from your checkpoint are actually the same to mine and they must not be ellipsoid or at least volumetric which is different from your demo video. Also your visualization looks different from the demo in that yours here show 2D flat and hexagon splats.

**Comment by jonathsch** [comment]
I added the Gaussian to the mesh visualization code I used to a new branch named `gaussian-viz`. Feel free to try it with your or my checkpoint. Regarding the different shape compared to the demo video, it is expected that we get different primitive shapes for different identities.
